manubot/pandoc/cite_filter.py

```diff
@@ -0,0 +1,219 @@
+            elem.content = elem.content[space_index + 1 :]
+        if len(elem.content) > 0 and type(elem.content[0]) == pf.SoftBreak:
+            elem.content.pop(0)
+
+
+def process_citations(doc):
+    """
+    Apply citation-by-identifier to a Python object representation of
+    Pandoc's Abstract Syntax Tree.
+
+    The following Pandoc metadata fields are considered:
+
+    - bibliography (use to define reference metadata manually)
+    - citekey-aliases (use to define tags for cite-by-id citations)
+    - manubot.requests-cache-path
+    - manubot.clear-requests-cache
+    - manubot.output-bibliography: path to write generated CSL JSON bibliography
+    """
+    citekey_aliases = doc.get_metadata("citekey-aliases", default={})
+    if not isinstance(citekey_aliases, dict):
+        logging.warning(
+            f"Expected metadata.citekey-aliases to be a dict, "
+            f"but received a {citekey_aliases.__class__.__name__}. Disregarding."
+        )
+        citekey_aliases = dict()
+
+    global_variables["citekey_aliases"] = citekey_aliases
+
+    manuscript_citekeys = set(global_variables["manuscript_citekeys"])
+
+    doc.walk(_get_reference_link_citekey_aliases)
+    doc.walk(_get_citekeys_action)
+    manuscript_citekeys = global_variables["manuscript_citekeys"]
+    manuscript_citekeys = sorted(
+        filter(
+            lambda x: is_valid_citekey(
+                x, allow_tag=True, allow_raw=True, allow_pandoc_xnos=True
+            ),
+            set(manuscript_citekeys),
+        )
+    )
+    global_variables["manuscript_citekeys"] = manuscript_citekeys
+    citekeys_df = get_citekeys_df(
+        manuscript_citekeys, global_variables["citekey_aliases"],
+    )
+    global_variables["citekeys_df"] = citekeys_df
+    global_variables["citekey_shortener"] = dict(
+        zip((citekeys_df["manuscript_citekey"]), citekeys_df["short_citekey"])
+    )
+    doc.walk(_citation_to_id_action)
+    manual_refs = doc.get_metadata("references", default=[])
+    bibliography_paths = doc.get_metadata("bibliography", default=[])
+    if not isinstance(bibliography_paths, list):
+        bibliography_paths = [bibliography_paths]
+    manual_refs = load_manual_references(
+        bibliography_paths, extra_csl_items=manual_refs
+    )
+    standard_citekeys = citekeys_df.standard_citekey.unique()
+    requests_cache_path = doc.get_metadata("manubot.requests-cache-path")
+    if requests_cache_path:
+        pathlib.Path(requests_cache_path).parent.mkdir(parents=True, exist_ok=True)
+    csl_items = generate_csl_items(
+        citekeys=standard_citekeys,
+        manual_refs=manual_refs,
+        requests_cache_path=doc.get_metadata("manubot.requests-cache-path"),
+        clear_requests_cache=doc.get_metadata("manubot.clear-requests-cache", False),
+    )
+    output_bibliography = doc.get_metadata("manubot.output-bibliography")
+    write_csl_json(csl_items, output_bibliography)
+    doc.metadata["bibliography"] = []
```

Comment:
What is this line doing?

Reply:
If the filter receives `metadata.bibliography` files, we load the references from those files. Therefore, we don't want pandoc-citeproc to re-read the references in those files. So we remove them from the metadata.